# PyTorch tensor functions

An short introduction about PyTorch and about the chosen functions. 

- torch.linspace()
- torch.trunc()
- torch.argmax()
- torch.trace()
- torch.abs()

Before we begin, let's install and import PyTorch

In [3]:
# Uncomment and run the appropriate command for your operating system, if required

# Linux / Binder
# !pip install numpy torch==1.7.0+cpu torchvision==0.8.1+cpu torchaudio==0.7.0 -f https://download.pytorch.org/whl/torch_stable.html

# Windows
!pip install numpy torch==1.7.0+cpu torchvision==0.8.1+cpu torchaudio==0.7.0 -f https://download.pytorch.org/whl/torch_stable.html

# MacOS
# !pip install numpy torch torchvision torchaudio

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in links: https://download.pytorch.org/whl/torch_stable.html
     |████████████████████████████████| 159.3 MB 3.5 kB/s 
     |████████████████████████████████| 11.8 MB 54.5 MB/s 
     |████████████████████████████████| 7.6 MB 8.4 MB/s 
  Attempting uninstall: torch
    Found existing installation: torch 1.11.0+cu113
    Uninstalling torch-1.11.0+cu113:
      Successfully uninstalled torch-1.11.0+cu113
  Attempting uninstall: torchvision
    Found existing installation: torchvision 0.12.0+cu113
    Uninstalling torchvision-0.12.0+cu113:
      Successfully uninstalled torchvision-0.12.0+cu113
  Attempting uninstall: torchaudio
    Found existing installation: torchaudio 0.11.0+cu113
    Uninstalling torchaudio-0.11.0+cu113:
      Successfully uninstalled torchaudio-0.11.0+cu113
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed.

In [4]:
# Import torch and other required modules
import torch

## Function 1 - torch.linspace 
Creates a one-dimensional tensor of size `steps` whose values are evenly spaced from `start` to `end`, inclusive. 

In [5]:
# Example 1 - working 
torch.linspace(start=-20, end=20, steps=6)

tensor([-20., -12.,  -4.,   4.,  12.,  20.])

As we can see, it returns a tensor whose first element is -20, last element 20 and creates 4 other middle elements of equal spacing.

In [7]:
# Example 2 - working
torch.linspace(start=0.5, end=5.5, steps=7)

tensor([0.5000, 1.3333, 2.1667, 3.0000, 3.8333, 4.6667, 5.5000])

Similarly, it returns a 7 element tensor equally spaced, starting with 0.5 and ending with 5.5. The elements can also be of `float` datatype

In [8]:
# Example 3 - breaking (to illustrate when it breaks)
torch.linspace(start=5, end =18, steps=0.2)

TypeError: ignored

As we can see, this doesn't execute because the argument `steps` should of type `int` (an integer). Because it essentialy also represents number of elements in the tensor.

In order to generate an equally spaced tensor with our required starting and ending elements, we can use the `torch.linspace()` function.


Let's save our work using Jovian before continuing.

In [9]:
!pip install jovian --upgrade --quiet

In [10]:
import jovian

In [11]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Please enter your API key ( from https://jovian.ai/ ):
API KEY: ··········
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/gauravnvn246/01-tensor-operations


'https://jovian.ai/gauravnvn246/01-tensor-operations'

## Function 2 - torch.trunc()

Returns a new tensor with the truncated integer values of the elements of `input`.

In [13]:
# Example 1 - working
a = torch.tensor([3.45, 4.67, 5.2231])
torch.trunc(a)

tensor([3., 4., 5.])

It takes a tensor as input, and returns the truncated integer value of the elements.

In [14]:
# Example 2 - working
b = torch.tensor([[1.2344, 3, 5.6739, 45.23],[23.56, 6.5, 9.99, 20.7]])
torch.trunc(b)

tensor([[ 1.,  3.,  5., 45.],
        [23.,  6.,  9., 20.]])

As we can see, it works with any dimensional tensor. The above illustrated example is that of a 2-D tensor.

In [15]:
# Example 3 - breaking (to illustrate when it breaks)
c = [4.55, 3.4, 6.7, 9.69]
torch.trunc(c)

TypeError: ignored

The `input` to the function must be a tensor, and the above example illustrates the case if the `input` was a `list`.

Closing comments about when to use this function

In [16]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/gauravnvn246/01-tensor-operations


'https://jovian.ai/gauravnvn246/01-tensor-operations'

## Function 3 - torch.argmax()

Returns the indices of the maximum value of all elements in the input tensor. If no additional argument is provided, it flattens the tensor and returns the modified index.

In [20]:
a = torch.randn(3,4)
a

tensor([[ 0.9195, -2.1946, -0.5625,  0.4970],
        [-1.1895,  0.2638,  1.4726, -2.1137],
        [ 0.0296,  0.0371, -1.5109, -0.4249]])

In [21]:
# Example 1 - working
torch.argmax(a)

tensor(6)

As we can see, the flattened tensor has the maximum value ***1.4726***, which is at index ***6***

In [22]:
# Example 2 - working
b = torch.randn(4,4)
b

tensor([[-2.0683,  0.5573,  2.3338, -1.5805],
        [-1.5211, -0.8713,  0.0827,  1.6446],
        [-0.2671, -1.2200, -1.1157,  0.7024],
        [ 1.7047, -0.6619, -2.5225,  1.0555]])

In [23]:
torch.argmax(b, dim=1)

tensor([2, 3, 3, 0])

The additional argument `dim` represents the dimension to reduce. If `None`, the argmax of the flattened input is returned. Here we can see that the indices of max value of each row is returned.

In [28]:
# Example 3 - breaking (to illustrate when it breaks)
c = torch.randn(6,7)
c

tensor([[ 0.3131,  1.0184, -0.0304, -0.2802, -0.2165, -0.9145,  0.6743],
        [ 0.4258,  0.6066,  2.0542, -2.0787,  0.2632, -0.4640,  1.0470],
        [ 1.3528,  0.3310, -0.0935, -1.1869,  1.9567,  0.4679,  0.0812],
        [-0.7887, -0.0187,  0.5269, -0.3270,  0.3688, -0.1145, -0.1475],
        [-1.5094, -0.2538, -1.1895, -0.0265, -0.2578,  1.0148,  0.2467],
        [ 0.1587, -0.0100, -1.0246,  0.7043, -1.1583,  0.6707, -0.8152]])

In [29]:
torch.argmax(c, dim=16)

IndexError: ignored

The argument `dim` has to be within range [-2,1] and cannot exceed it. This is because it doesn't make sense to reduce it to a dimension out of this range

This function can be used in Deep Learning models in the final layer of the fully connected layers to return the predicted output, since that would have the highest probability.

In [30]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/gauravnvn246/01-tensor-operations


'https://jovian.ai/gauravnvn246/01-tensor-operations'

## Function 4 - torch.trace()
Returns the sum of the elements of the diagonal of the input 2-D matrix.



In [32]:
# Example 1 - working
a = torch.arange(12,21).view(3,3)
a

tensor([[12, 13, 14],
        [15, 16, 17],
        [18, 19, 20]])

In [33]:
torch.trace(a)

tensor(48)

Returns the trace of the 2-D matrix, i.e sum of diagonal elements of the matrix.

In [34]:
# Example 2 - working
b = torch.ones(4,5)
b

tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])

In [35]:
torch.trace(b)

tensor(4.)

It need not be a square matrix, but it has to be a 2-D matrix.

In [36]:
# Example 3 - breaking (to illustrate when it breaks)
c = torch.randn(3,4,5)
c

tensor([[[-0.2878, -0.9097,  0.3620, -0.2601,  0.5944],
         [-0.8498, -0.8754,  1.1506, -1.7598,  0.8556],
         [ 0.4250, -0.4767,  0.7335, -0.1357,  0.0765],
         [ 0.7472, -1.1382,  0.9754, -0.1346, -0.4950]],

        [[ 1.0426,  0.7023, -0.4634, -0.9011,  2.5019],
         [-0.5604,  0.4726, -0.4751,  0.9597, -0.3263],
         [ 0.1839, -2.1962,  0.1732,  1.5222,  0.4425],
         [-1.2589, -0.9544,  1.4109, -0.3782, -0.1629]],

        [[ 0.0046, -2.1200,  1.3046, -0.0171, -0.0109],
         [ 1.1178,  0.6551, -0.9698, -1.1991,  0.4692],
         [ 0.1873,  1.0312, -0.0096,  0.4838, -1.0428],
         [-1.1848, -0.6585, -0.3818, -0.5279,  1.0622]]])

In [37]:
torch.trace(c)

RuntimeError: ignored

As said before, this matrix is 3-D. Therefore it doesn't make sense for trace of a 3-D matrix. Hence the function `torch.trace` doesn't work.

In questions involving linear algebra, this function would be useful.

In [38]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/gauravnvn246/01-tensor-operations


'https://jovian.ai/gauravnvn246/01-tensor-operations'

## Function 5 - torch.abs()

Computes the absolute value of each element in `input`.



In [55]:
# Example 1 - working
a = torch.arange(-3,9).view(3,4)
a

tensor([[-3, -2, -1,  0],
        [ 1,  2,  3,  4],
        [ 5,  6,  7,  8]])

In [56]:
torch.abs(a)

tensor([[3, 2, 1, 0],
        [1, 2, 3, 4],
        [5, 6, 7, 8]])

As we can see, it returns the absolute value of all the elements in the tensor.

In [57]:
# Example 2 - working
b = torch.randn(3,2)
b

tensor([[ 3.6929, -0.3092],
        [-0.2979,  0.5679],
        [-1.0185, -0.7504]])

In [58]:
torch.abs(b)

tensor([[3.6929, 0.3092],
        [0.2979, 0.5679],
        [1.0185, 0.7504]])

It also works in any dimension tensor, with any numeric datatype.

In [59]:
# Example 3 - breaking (to illustrate when it breaks)
torch.abs(torch.tensor([[1, -2], [-3, 4, -5]]))

ValueError: ignored

Function `torch.abs` is going to work fine always except a case when you do not pass the correct tensor.

We can use this function when you need a tensor having all values positive but the tensor that you have may have some negative numbers.

In [60]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/gauravnvn246/01-tensor-operations


'https://jovian.ai/gauravnvn246/01-tensor-operations'

## Conclusion

We have explored the Pytorch documentation and have tried out different tensor operations.

## Reference Links
Provide links to your references and other interesting articles about tensors
* Official documentation for tensor operations: https://pytorch.org/docs/stable/torch.html
* ...

In [ ]:
jovian.commit(project='01-tensor-operations')

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
